src/lsm/set_associative_cache.zig

```diff
@@ -0,0 +1,423 @@
+            value_ptr: *align(value_alignment) Value,
+            found_existing: bool,
+        };
+
+        pub fn get_or_put(self: *Self, key: Key) GetOrPutResult {
+            return self.get_or_put_preserve_locked(
+                void,
+                struct {
+                    inline fn locked(_: void, _: *const Value) bool {
+                        return false;
+                    }
+                }.locked,
+                {},
+                key,
+            );
+        }
+
+        /// Get a pointer to an existing value with the given key if the key is in the cache.
+        /// If the key is not in the cache, add it to the cache, evicting older entries if needed.
+        /// Never evicts keys for which locked() returns true.
+        /// The caller must guarantee that locked() returns true for less than layout.ways keys.
+        pub fn get_or_put_preserve_locked(
+            self: *Self,
+            comptime Context: type,
+            comptime locked: fn (Context, *align(value_alignment) const Value) callconv(.Inline) bool,
+            context: Context,
+            key: Key,
+        ) GetOrPutResult {
+            const set = self.associate(key);
+
+            if (self.search(set, key)) |existing| {
+                assert(!locked(context, existing));
+                return .{
+                    .value_ptr = existing,
+                    .found_existing = true,
+                };
+            }
+
+            const clock_index = @divExact(set.offset, layout.ways);
+
+            const max_iterations = layout.ways * math.maxInt(Count);
+            var safety_counter: usize = 0;
+
+            var way = Clocks.get(self.clocks, clock_index);
+            while (safety_counter < max_iterations + 1) : ({
+                way +%= 1;
+                safety_counter += 1;
+            }) {
+                // We pass a value pointer to the callback here so that a cache miss
+                // can be avoided if the caller is able to determine if the value is
+                // locked by comparing pointers directly.
+                if (locked(context, @alignCast(value_alignment, &set.values[way]))) continue;
+
+                const count = Counts.get(self.counts, set.offset + way);
+
+                // Free way found
+                if (count == 0) break;
+
+                Counts.set(self.counts, set.offset + way, count - 1);
+            }
+            assert(safety_counter < max_iterations);
+
+            Clocks.set(self.clocks, clock_index, way +% 1);
+
+            set.tags[way] = set.tag;
+
+            Counts.set(self.counts, set.offset + way, 1);
+
+            return .{
+                .value_ptr = @alignCast(value_alignment, &set.values[way]),
+                .found_existing = false,
+            };
+        }
+
+        fn search(self: *Self, set: Set, key: Key) ?*align(value_alignment) Value {
+            for (set.tags) |tag, way| {
+                // TODO self.counts.get()
```

Comment:
I didn't write this TODO, but I think it's about changing the `PackedUnsignedIntegerArray` type to store the slice so that we could store use `self.counts.get()` instead of `Counts.get(self.counts)` which seems like it would be nice to do sometime.